File: rust/libipc/src/chunk_storage.rs
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception OR MIT
// SPDX-FileCopyrightText: 2025-2026 natyamatsya and thoth-ipc contributors
//
// Port of cpp-ipc/src/libipc/ipc.cpp: chunk_info_t / id_pool / acquire_storage /
// find_storage / recycle_storage. **Byte-exact with the C++ chunk storage** so a
// C++ sender's large (>64B) messages can be read by a Rust receiver — see
// context/xlang-channel-abi.md §6c.
//
// A C++ sender stores messages >large_msg_limit (64B) in a chunk shm and pushes a
// single msg_t with storage_=true and the storage_id in the payload. A Rust
// receiver reads it via find_storage and frees it via recycle_storage. (A Rust
// sender keeps fragmenting instead — C++ recv reassembles — so acquire_storage is
// present for symmetry but unused by the current send path.)
//
// Chunk shm layout for a given `chunk_size` (name __IPC_SHM__CHUNK_INFO__<size>):
//   [ chunk_info_t (40B) ] [ chunk_t of chunk_size bytes ] × MAX_COUNT
// chunk_info_t: id_pool { next_[32]; cursor_; prepared_ } + spin_lock @36.
// chunk_t: conns (AtomicU32) @0, payload @ make_align(8,4)=8.

#![allow(dead_code)] // acquire_storage/id_pool helpers unused by the fragmenting send path

use std::io;
use std::sync::atomic::{AtomicU32, Ordering};
use std::sync::Arc;

use crate::abi_generated as abi;

#[cfg(unix)]
use crate::platform::posix::{cached_shm_acquire, cached_shm_purge, cached_shm_release, CachedShm};
use crate::shm::ShmHandle;
#[cfg(not(unix))]
use crate::shm::ShmOpenMode;

/// Max large-message slots per chunk size (C++ `large_msg_cache = 32`).
pub const MAX_COUNT: usize = abi::large_msg_cache;
/// Chunk-size alignment (C++ `large_msg_align = 1024`).
pub const CHUNK_ALIGN: usize = abi::large_msg_align;
/// Per-chunk header = `make_align(alignof(max_align_t)=8, sizeof(atomic<cc_t>)=4)` = 8.
const CHUNK_HEADER: usize = abi::chunk_header_size;

/// A `storage_id` (C++ `storage_id_t = int32`); < 0 means invalid.
pub type StorageId = i32;

// ---------------------------------------------------------------------------
// chunk_info_t — byte-exact with C++ { id_pool pool_; spin_lock lock_; }
// ---------------------------------------------------------------------------

#[repr(C)]
struct ChunkInfo {
    next_: [u8; MAX_COUNT], // @0  id_pool free-list links
    cursor_: u8,            // @32 head of the free list
    prepared_: u8,          // @33 id_pool::prepared_ (bool)
    _pad: [u8; 2],          // @34..36
    #[cfg(target_vendor = "apple")]
    lock_: libc::os_unfair_lock, // @36 (C++ spin_lock)
    #[cfg(not(target_vendor = "apple"))]
    lock_: AtomicU32, // @36 (C++ generic spin_lock = atomic<u32> TAS-spin)
}

const _: () = {
    assert!(std::mem::size_of::<ChunkInfo>() == abi::chunk_info_size);
    assert!(std::mem::offset_of!(ChunkInfo, next_) == 0);
    assert!(std::mem::offset_of!(ChunkInfo, cursor_) == 32);
    assert!(std::mem::offset_of!(ChunkInfo, prepared_) == 33);
    assert!(std::mem::offset_of!(ChunkInfo, lock_) == 36);
};

impl ChunkInfo {
    /// Total shm size: header + MAX_COUNT chunks of `chunk_size` bytes.
    pub const fn shm_size(chunk_size: usize) -> usize {
        std::mem::size_of::<ChunkInfo>() + MAX_COUNT * chunk_size
    }

    /// C++ id_pool::prepare()/init(): a fresh (zeroed) pool is "invalid" → build the
    /// free list `next_[i] = i+1`. Call under the lock.
    fn prepare(&mut self) {
        if self.prepared_ == 0 && self.cursor_ == 0 && self.next_[0] == 0 {
            for i in 0..MAX_COUNT {
                self.next_[i] = (i + 1) as u8;
            }
        }
        self.prepared_ = 1;
    }

    /// C++ id_pool::acquire(): id = cursor_; cursor_ = next_[id].
    fn acquire(&mut self) -> StorageId {
        if self.cursor_ as usize >= MAX_COUNT {
            return -1;
        }
        let id = self.cursor_ as StorageId;
        self.cursor_ = self.next_[id as usize];
        id
    }

    /// C++ id_pool::release(): next_[id] = cursor_; cursor_ = id.
    fn release(&mut self, id: StorageId) {
        if id < 0 || id as usize >= MAX_COUNT {
            return;
        }
        self.next_[id as usize] = self.cursor_;
        self.cursor_ = id as u8;
    }
}

/// Lock the chunk_info_t spin_lock: os_unfair_lock on Apple; on other targets the
/// C++ generic spin_lock — an atomic<u32> test-and-set spin (1 = locked, 0 = free),
/// byte-exact at lock_ @36 so a C++ peer and this port serialise pool access.
unsafe fn chunk_lock(info: &ChunkInfo) {
    #[cfg(target_vendor = "apple")]
    {
        libc::os_unfair_lock_lock(&info.lock_ as *const _ as *mut libc::os_unfair_lock);
    }
    #[cfg(not(target_vendor = "apple"))]
    {
        let mut k = 0u32;
        while info.lock_.swap(1, Ordering::Acquire) != 0 {
            crate::spin_lock::adaptive_yield_pub(&mut k);
        }
    }
}

unsafe fn chunk_unlock(info: &ChunkInfo) {
    #[cfg(target_vendor = "apple")]
    {
        libc::os_unfair_lock_unlock(&info.lock_ as *const _ as *mut libc::os_unfair_lock);
    }
    #[cfg(not(target_vendor = "apple"))]
    {
        info.lock_.store(0, Ordering::Release);
    }
}

// ---------------------------------------------------------------------------
// Chunk-size calculation — byte-exact with C++ calc_chunk_size
// ---------------------------------------------------------------------------

/// `calc_chunk_size(size) = ceil((CHUNK_HEADER + size) / CHUNK_ALIGN) * CHUNK_ALIGN`
/// (C++: make_align(8, align_chunk_size(make_align(8, sizeof(atomic<cc_t>)) + size))).
/// `size` is the message size. The chunk-shm name embeds this, so it must match C++.
pub fn calc_chunk_size(size: usize) -> usize {
    let x = CHUNK_HEADER + size;
    x.div_ceil(CHUNK_ALIGN) * CHUNK_ALIGN
}

// ---------------------------------------------------------------------------
// Process-local chunk shm cache
// ---------------------------------------------------------------------------

/// A cached chunk shm handle — wraps a `CachedShm` so same-process endpoints
/// share the same mmap (required for data coherency on macOS with MAP_SHARED).
#[cfg(unix)]
pub struct ChunkShmHandle {
    cached: Arc<CachedShm>,
    name: String,
}

#[cfg(unix)]
impl ChunkShmHandle {
    pub fn get(&self) -> *mut u8 {
        self.cached.shm.as_mut_ptr()
    }
}

#[cfg(unix)]
impl Drop for ChunkShmHandle {
    fn drop(&mut self) {
        cached_shm_release(chunk_cache(), &self.name);
    }
}

#[cfg(unix)]
fn chunk_cache() -> &'static std::sync::Mutex<crate::platform::posix::ShmCache> {
    use std::sync::{Mutex, OnceLock};
    static CACHE: OnceLock<Mutex<crate::platform::posix::ShmCache>> = OnceLock::new();
    CACHE.get_or_init(|| Mutex::new(crate::platform::posix::ShmCache::new()))
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/// Chunk-shm name, byte-exact with C++ make_prefix(prefix, "CHUNK_INFO__", chunk_size).
/// `full_prefix` must be the prefix-global `"{prefix}__IPC_SHM__"` (NO channel name).
fn chunk_shm_name(full_prefix: &str, chunk_size: usize) -> String {
    format!("{full_prefix}CHUNK_INFO__{chunk_size}")
}

/// Open (or create) the chunk-storage shm for `chunk_size`-byte chunks.
#[cfg(unix)]
pub fn open_chunk_shm(full_prefix: &str, chunk_size: usize) -> io::Result<ChunkShmHandle> {
    let name = chunk_shm_name(full_prefix, chunk_size);
    let size = ChunkInfo::shm_size(chunk_size);
    let cached = cached_shm_acquire(chunk_cache(), &name, size, |_| Ok(()))?;
    Ok(ChunkShmHandle { cached, name })
}

#[cfg(not(unix))]
pub fn open_chunk_shm(full_prefix: &str, chunk_size: usize) -> io::Result<ShmHandle> {
    let name = chunk_shm_name(full_prefix, chunk_size);
    let size = ChunkInfo::shm_size(chunk_size);
    ShmHandle::acquire(&name, size, ShmOpenMode::CreateOrOpen)
}

/// C++ acquire_storage: allocate a chunk id, stamp its conns bitmask, return the
/// payload pointer. (Unused by the fragmenting send path; kept for symmetry.)
pub fn acquire_storage(base: *mut u8, chunk_size: usize, conns: u32) -> Option<(StorageId, *mut u8)> {
    let info = unsafe { &mut *(base as *mut ChunkInfo) };
    unsafe { chunk_lock(info) };
    info.prepare();
    let id = info.acquire();
    unsafe { chunk_unlock(info) };
    if id < 0 {
        return None;
    }
    let conns_ptr = unsafe { chunk_conns_ptr(base, chunk_size, id) };
    unsafe { (*conns_ptr).store(conns, Ordering::Relaxed) };
    Some((id, chunk_payload_ptr(base, chunk_size, id)))
}

/// C++ find_storage: pointer to the payload of chunk `id` (offset CHUNK_HEADER).
pub fn find_storage(base: *mut u8, chunk_size: usize, id: StorageId) -> Option<*mut u8> {
    if id < 0 || id as usize >= MAX_COUNT {
        return None;
    }
    Some(chunk_payload_ptr(base, chunk_size, id))
}

/// C++ recycle_storage / sub_rc<broadcast>: clear this receiver's bit from the
/// chunk's conns; when it reaches 0 (last reader), release the id to the pool.
pub fn recycle_storage(base: *mut u8, chunk_size: usize, id: StorageId, conn_id: u32) {
    if id < 0 || id as usize >= MAX_COUNT {
        return;
    }
    let conns = unsafe { &*chunk_conns_ptr(base, chunk_size, id) };
    let mut k = 0u32;
    let last = loop {
        let cur = conns.load(Ordering::Acquire);
        let nxt = cur & !conn_id;
        if conns
            .compare_exchange_weak(cur, nxt, Ordering::Release, Ordering::Relaxed)
            .is_ok()
        {
            break nxt == 0;
        }
        crate::spin_lock::adaptive_yield_pub(&mut k);
    };
    if last {
        let info = unsafe { &mut *(base as *mut ChunkInfo) };
        unsafe { chunk_lock(info) };
        info.release(id);
        unsafe { chunk_unlock(info) };
    }
}

/// Remove the chunk-storage shm segment for `chunk_size`.
pub fn clear_chunk_shm(full_prefix: &str, chunk_size: usize) {
    let name = chunk_shm_name(full_prefix, chunk_size);
    #[cfg(unix)]
    cached_shm_purge(chunk_cache(), &name);
    ShmHandle::clear_storage(&name);
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

/// Pointer to chunk `id`'s conns bitmask (AtomicU32 @ start of the chunk).
unsafe fn chunk_conns_ptr(base: *mut u8, chunk_size: usize, id: StorageId) -> *mut AtomicU32 {
    let chunks_base = base.add(std::mem::size_of::<ChunkInfo>());
    chunks_base.add(chunk_size * id as usize) as *mut AtomicU32
}

/// Pointer to chunk `id`'s payload (after the CHUNK_HEADER-byte conns header).
fn chunk_payload_ptr(base: *mut u8, chunk_size: usize, id: StorageId) -> *mut u8 {
    unsafe {
        let chunks_base = base.add(std::mem::size_of::<ChunkInfo>());
        chunks_base.add(chunk_size * id as usize).add(CHUNK_HEADER)
    }
}

File: swift/libipc/Sources/LibIPC/Transport/ChunkStorage.swift
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2025-2026 natyamatsya contributors
//
// Port of cpp-ipc/src/libipc/ipc.cpp: chunk_info_t / acquire_storage /
// find_storage / recycle_storage / id_pool.
//
// Large messages (> DATA_LENGTH bytes) are stored in a separate named shm
// segment instead of being fragmented across multiple ring slots. Only a
// 4-byte `storageId` is placed in the ring slot's data field.
//
// Shared-memory layout for a given `chunkSize`:
//
//   [ ChunkInfo header ]
//   [ chunkSize bytes ] × MAX_COUNT   ← chunk data array
//
// ChunkInfo header (at offset 0):
//   lock    : UInt32   (spin-lock: 0=free, 1=held)
//   cursor  : UInt8    (head of the free-list)
//   next    : [UInt8; MAX_COUNT]  (free-list links)
//
// Each chunk (chunkSize bytes):
//   conns   : UInt32   (broadcast connection bitmask — ref-count per receiver)
//   payload : [UInt8; chunkSize - CHUNK_HEADER]

import Darwin.POSIX
import Atomics

// MARK: - Constants

/// Maximum number of large-message slots per chunk size (matches C++ `large_msg_cache = 32`).
let chunkMaxCount: Int = 32

/// Alignment for chunk sizes (matches C++ `large_msg_align = 1024`).
let chunkAlign: Int = 1024

/// Bytes consumed by the per-chunk connection bitmask at the start of each chunk.
let chunkHeaderSize: Int = MemoryLayout<UInt32>.size  // 4

/// A storage slot identifier; -1 means invalid / not allocated.
typealias StorageId = Int32

// MARK: - Chunk-size calculation

/// Round `payloadSize` up to the next multiple of `chunkAlign`, add the per-chunk
/// header, then align the total to 16 bytes.
/// Mirrors C++ `calc_chunk_size` / Rust `calc_chunk_size`.
func calcChunkSize(_ payloadSize: Int) -> Int {
    let aligned = ((payloadSize + chunkAlign - 1) / chunkAlign) * chunkAlign
    let total = MemoryLayout<UInt32>.size + aligned
    let align = 16  // MemoryLayout<(UInt64, UInt64)>.alignment
    return (total + align - 1) / align * align
}

// MARK: - ChunkInfo layout helpers

/// Offset of the spin-lock field within ChunkInfo.
private let ciLockOffset = 0
/// Offset of the cursor field within ChunkInfo.
private let ciCursorOffset = MemoryLayout<UInt32>.size  // 4
/// Offset of the next[] array within ChunkInfo.
private let ciNextOffset = ciCursorOffset + 1  // 5
/// Total size of the ChunkInfo header.
let chunkInfoSize: Int = ciNextOffset + chunkMaxCount  // 5 + 32 = 37, but we round up

// We store ChunkInfo as a flat byte region; use helpers to access fields.

private func ciLockPtr(_ base: UnsafeMutableRawPointer) -> UnsafeMutablePointer<UInt32> {
    base.advanced(by: ciLockOffset).assumingMemoryBound(to: UInt32.self)
}

private func ciCursorPtr(_ base: UnsafeMutableRawPointer) -> UnsafeMutablePointer<UInt8> {
    base.advanced(by: ciCursorOffset).assumingMemoryBound(to: UInt8.self)
}

private func ciNextPtr(_ base: UnsafeMutableRawPointer) -> UnsafeMutablePointer<UInt8> {
    base.advanced(by: ciNextOffset).assumingMemoryBound(to: UInt8.self)
}

/// Actual header size rounded up to 16-byte alignment so chunk array starts aligned.
let chunkInfoAlignedSize: Int = (ciNextOffset + chunkMaxCount + 15) / 16 * 16  // = 48

/// Total shm size for a chunk-storage segment with `chunkSize`-byte chunks.
func chunkShmSize(_ chunkSize: Int) -> Int {
    chunkInfoAlignedSize + chunkMaxCount * chunkSize
}

// MARK: - Spin-lock helpers (raw UInt32 in shm)

private func rawSpinLock(_ ptr: UnsafeMutablePointer<UInt32>) {
    var k: UInt32 = 0
    ptr.withMemoryRebound(to: UInt32.AtomicRepresentation.self, capacity: 1) { rep in
        while true {
            var expected: UInt32 = 0
            let (exchanged, _) = UInt32.AtomicRepresentation.atomicWeakCompareExchange(
                expected: expected,
                desired: 1,
                at: rep,
                successOrdering: .acquiring,
                failureOrdering: .relaxed
            )
            _ = expected  // suppress unused warning
            if exchanged { return }
            adaptiveYieldSync(&k)
        }
    }
}

private func rawSpinUnlock(_ ptr: UnsafeMutablePointer<UInt32>) {
    ptr.withMemoryRebound(to: UInt32.AtomicRepresentation.self, capacity: 1) { rep in
        UInt32.AtomicRepresentation.atomicStore(0, at: rep, ordering: .releasing)
    }
}

// MARK: - Free-list init / acquire / release

/// Initialise the free-list if the pool looks uninitialised (all-zero = fresh shm).
/// Must be called while the spin-lock is held.
private func ensureInit(_ base: UnsafeMutableRawPointer) {
    let cursor = ciCursorPtr(base).pointee
    let next0 = ciNextPtr(base).pointee
    guard cursor == 0 && next0 == 0 else { return }
    let nextArr = ciNextPtr(base)
    for i in 0..<chunkMaxCount {
        nextArr.advanced(by: i).pointee = UInt8(i + 1)
    }
    // cursor stays 0 — first free slot is index 0
}

private func chunkAcquire(_ base: UnsafeMutableRawPointer) -> StorageId {
    let cursor = ciCursorPtr(base).pointee
    guard cursor < UInt8(chunkMaxCount) else { return -1 }
    let id = StorageId(cursor)
    ciCursorPtr(base).pointee = ciNextPtr(base).advanced(by: Int(id)).pointee
    return id
}

private func chunkRelease(_ base: UnsafeMutableRawPointer, id: StorageId) {
    guard id >= 0 && id < StorageId(chunkMaxCount) else { return }
    ciNextPtr(base).advanced(by: Int(id)).pointee = ciCursorPtr(base).pointee
    ciCursorPtr(base).pointee = UInt8(id)
}

// MARK: - Chunk pointer helpers

/// Pointer to the `UInt32` connection bitmask at the start of chunk `id`.
func chunkConnsPtr(
    _ base: UnsafeMutableRawPointer,
    chunkSize: Int,
    id: StorageId
) -> UnsafeMutablePointer<UInt32> {
    base.advanced(by: chunkInfoAlignedSize + chunkSize * Int(id))
        .assumingMemoryBound(to: UInt32.self)
}

/// Pointer to the payload bytes of chunk `id` (after the 4-byte conn header).
func chunkPayloadPtr(
    _ base: UnsafeMutableRawPointer,
    chunkSize: Int,
    id: StorageId
) -> UnsafeMutableRawPointer {
    base.advanced(by: chunkInfoAlignedSize + chunkSize * Int(id) + chunkHeaderSize)
}

// MARK: - Public API

/// Open (or create) the chunk-storage shm segment for `chunkSize`-byte chunks.
func openChunkShm(prefix: String, chunkSize: Int) throws(IpcError) -> ShmHandle {
    let name = "\(prefix)CH_CONN__\(chunkSize)"
    return try ShmHandle.acquire(name: name, size: chunkShmSize(chunkSize), mode: .createOrOpen)
}

/// Acquire a free slot from the chunk-storage shm.
///
/// Returns `(storageId, payloadPointer)` on success, or `nil` if the pool is exhausted.
/// Mirrors C++ `acquire_storage` / Rust `acquire_storage`.
func acquireStorage(
    shm: borrowing ShmHandle,
    chunkSize: Int,
    conns: UInt32
) -> (StorageId, UnsafeMutableRawPointer)? {
    let base = shm.ptr
    let lockPtr = ciLockPtr(base)
    rawSpinLock(lockPtr)
    ensureInit(base)
    let id = chunkAcquire(base)
    rawSpinUnlock(lockPtr)
    guard id >= 0 else { return nil }
    let payload = chunkPayloadPtr(base, chunkSize: chunkSize, id: id)
    // Store the connection bitmask in the per-chunk header.
    chunkConnsPtr(base, chunkSize: chunkSize, id: id).withMemoryRebound(
        to: UInt32.AtomicRepresentation.self, capacity: 1
    ) { rep in
        UInt32.AtomicRepresentation.atomicStore(conns, at: rep, ordering: .relaxed)
    }
    return (id, payload)
}

/// Return a pointer to the payload of chunk `id`.
/// Mirrors C++ `chunk_info_t::at(chunk_size, id)->data()` / Rust `find_storage`.
func findStorage(
    shm: borrowing ShmHandle,
    chunkSize: Int,
    id: StorageId
) -> UnsafeMutableRawPointer? {
    guard id >= 0 && id < StorageId(chunkMaxCount) else { return nil }
    return chunkPayloadPtr(shm.ptr, chunkSize: chunkSize, id: id)
}

/// Clear the receiver's bit from the chunk's connection bitmask.
/// When the bitmask reaches zero (last reader), release the slot back to the pool.
/// Mirrors C++ `recycle_storage` / Rust `recycle_storage`.
func recycleStorage(
    shm: borrowing ShmHandle,
    chunkSize: Int,
    id: StorageId,
    connId: UInt32
) {
    guard id >= 0 && id < StorageId(chunkMaxCount) else { return }
    let base = shm.ptr
    let connsRaw = chunkConnsPtr(base, chunkSize: chunkSize, id: id)
    var k: UInt32 = 0
    var isLast = false
    connsRaw.withMemoryRebound(
        to: UInt32.AtomicRepresentation.self, capacity: 1
    ) { rep in
        while true {
            let cur = UInt32.AtomicRepresentation.atomicLoad(at: rep, ordering: .acquiring)
            let nxt = cur & ~connId
            let (didExchange, _) = UInt32.AtomicRepresentation.atomicWeakCompareExchange(
                expected: cur,
                desired: nxt,
                at: rep,
                successOrdering: .releasing,
                failureOrdering: .relaxed
            )
            if didExchange { isLast = (nxt == 0); return }
            adaptiveYieldSync(&k)
        }
    }
    if isLast {
        let lockPtr = ciLockPtr(base)
        rawSpinLock(lockPtr)
        chunkRelease(base, id: id)
        rawSpinUnlock(lockPtr)
    }
}

/// Remove the chunk-storage shm segment for `chunkSize`.
func clearChunkShm(prefix: String, chunkSize: Int) {
    let name = "\(prefix)CH_CONN__\(chunkSize)"
    ShmHandle.clearStorage(name: name)
}
